### PR TITLE
BLUEBUTTON-1086: first draft of new Jenkinsfile for monorepo

### PR DESCRIPTION
**Pull request by karlmdavis**
This is a first draft of our new build process, which hasn't been tested yet. I'm likely going to merge it before doing that testing, so that other folks can start on work that needs to integrate with the approach and code. _Hopefully_ there won't be any major changes required, but... 🤷‍♂.

Please see the other closed PRs for more background & history on what I've got in-flight. They tend to contain massive `git mv`s in some of their commits, so you'll probably want to look at the individual commits for a better sense of what was actually changed:
* #1
* #2
* #3

There's still quite a bit more work to do before the monorepo migration is "done", including:

* Rename all of the Java packages and classes to replace references to "Blue Button" with "BFD".
* Rename all of the variables, similarly.
* Update the various `README.md`s and other documentation to reflect the new structure and naming.
* Update all of the Ansible roles and playbooks similarly: names, variables, etc.

And perhaps most challengingly:

* Update all of the server paths, files, etc. to reflect the new names.
    * Will likely wait on this until after the CCS migration is complete, as it's much simpler to do in an immutable deploy scenario (don't have to move anything on existing servers).

Nonetheless, once I merge the changes represented by this particular PR, I think the migration will be "done enough" for most ops work to proceed. The external interfaces to most things should remain stable _enough_ for that to be reasonable.

https://jira.cms.gov/browse/BLUEBUTTON-1086

### CODE REVIEW

File: Jenkinsfile [hunk 1]
```diff
@@ -0,0 +1,201 @@
...
+	lock(resource: 'env_test', inversePrecendence: true) {
+		milestone(label: 'stage_deploy_test_start')
+
+		node {
+			scriptForDeploys.deploy('test', platinumAmiId, appsBuildResult)
+		}
+	}
+}
+
+stage('Manual Approval') {
```
[the remark this answers is not in the document]
Comment (jzulim) [reply Jenkinsfile]: Dropped this in the backlog: https://jira.cms.gov/browse/BLUEBUTTON-1246

File: ops/ansible/playbooks-healthapt/backend.yml [hunk 2]
```diff
@@ -99,8 +97,7 @@
         data_server_appserver_management_username: "{{ vault_data_server_appserver_management_username }}"
         data_server_appserver_management_password: "{{ vault_data_server_appserver_management_password }}"
         data_server_appserver_https_port: "{{ vault_data_server_appserver_https_port }}"
-        data_server_war_name: "bfd-server-war-{{ data_server_version }}.war"
-        data_server_war_local_dir: "~/.m2/repository/gov/cms/bfd/bfd-server-war/{{ data_server_version }}"
+        # Note: the `data_server_war` variable should have been provided via `--extra-vars` on the command line.
```
Comment (jzulim) [inline ops/ansible/playbooks-healthapt/backend.yml]: noted, will need update packer ansible provisioner for pipeline and server to include these.

File: ops/ansible/roles/bfd-pipeline/tasks/main.yml [hunk 3]
```diff
@@ -45,62 +45,17 @@
   become: true
   when: data_pipeline_tmp_dir is defined and stat_tmp.stat.exists == false
 
-# If the app's JAR is in S3, grab it from there.
```
Comment (jzulim) [inline ops/ansible/roles/bfd-pipeline/tasks/main.yml]: Removing this is great, as in CCS it's taken care of in platinum build.

File: ops/ansible/roles/bfd-server/README.md [hunk 4]
```diff
@@ -17,13 +17,9 @@ This role is highly configurable, though it tries to provide reasonable defaults
...
 
-* `data_server_artifacts_mode`: Set to "`s3`" if the WAR and app server should be copied from an AWS S3 bucket or "`local`" if they should be copied from a path local to the Ansible management host.
-* `data_server_artifacts_s3_bucket`: The name of the S3 bucket containing the artifacts to be deployed. Must be set if `data_server_artifacts_mode` is "`s3`".
-* `data_server_appserver_installer_name`: The name of the app server installation bundle to be deployed.
-* `data_server_appserver_name`: The name of the directory that will be unzipped by the app server installation bundle, e.g. "`wildfly-8.1.0.Final`".
-* `data_server_appserver_local_dir`: The local directory for the app server installation bundle to be deployed. Must be set if `data_server_artifacts_mode` is "`local`".
-* `data_server_war_name`: The name of the WAR file to be deployed (as saved in either S3 or a local directory).
-* `data_server_war_local_dir`: The local directory of the WAR file to be deployed. Must be set if `data_server_artifacts_mode` is "`local`".
+* `data_server_container`: The path (on the management system) to the app server installation bundle to be deployed.
```
Comment (jzulim) [inline ops/ansible/roles/bfd-server/README.md]: I'm assuming we'll need to add this to the jenkins instance(s). This it? https://download.jboss.org/wildfly/8.1.0.Final/wildfly-8.1.0.Final.zip

**Comment by jzulim** [comment]
I gave this a line by line read. The changes make total sense to me and I feel I can pick this up and build out the groovy scripts for Platinum Build and Terraform Deploy stages. Thanks for putting this all together Karl, it's REALLY helpful and has helped me with a vision / path to move forward with the build and deployment stages.

File: ops/ansible/roles/bfd-server/README.md (hunk 4 above)
Comment (karlmdavis) [reply ops/ansible/roles/bfd-server/README.md]: Nope, not quite. Take a look at `build.groovy` and note the `BuildResults` struct it returns. That struct contains the paths to the JARs, WARs, and Wildfly installer. The struct will get passed into the AMI build, so all those artifacts can be used there.

(I just noticed that it _isn't_ being passed in right now, so I'll fix that.)

File: Jenkinsfile (hunk 1 above)
[the remark this answers is not in the document]
Comment (karlmdavis) [reply Jenkinsfile]: ty!

File: apps/build.groovy [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+			]
+	) {
+		sh "${mvnHome}/bin/mvn --settings $MAVEN_SETTINGS --toolchains $MAVEN_TOOLCHAINS ${args}"
+	}
+}
+
+/**
+ * Models the results of a call to {@link #build}: contains the paths to the artifacts that were built.
+ */
+class AppBuildResults implements Serializable {
```
Comment (karlmdavis) [inline apps/build.groovy]: Just a note: I _think_ I can get away with embedding a class here, but I'm not entirely sure. Won't know either way until I actually try running this and see if/how it all blows up.

I might have to move these embedded classes (this and `AmiIds`) up to the top-level script, or worst-case replace them with a hokey `String[]` array.

File: ops/ansible/roles/bfd-pipeline/tasks/main.yml (hunk 3 above)
Comment (karlmdavis) [reply ops/ansible/roles/bfd-pipeline/tasks/main.yml]: Yeah. Only the old ETL benchmarks were using it, and I'd rather rework them to use AMIs (if I ever need them again).

File: ops/ansible/playbooks-healthapt/backend.yml (hunk 2 above)
Comment (karlmdavis) [reply ops/ansible/playbooks-healthapt/backend.yml]: Please note that these paths will be passed in via the `AppBuildResults` instance that gets passed to the Groovy method.

File: Jenkinsfile [hunk 6]
```diff
@@ -0,0 +1,201 @@
...
+	}
+}
+
+stage('Deploy to hhsdevcloud') {
+	if (willDeployToProdEnvs && params.deploy_hhsdevcloud) {
+		lock(resource: 'env_hhsdevcloud', inversePrecendence: true) {
+			milestone(label: 'stage_deploy_hhsdevcloud')
+
+			node {
+				scriptForDeploysHhsdevcloud.deploy(appsBuildResult)
```
[the remark this answers is not in the document]
Comment (karlmdavis) [reply Jenkinsfile]: Eh, not worth worrying about: deploys to the environment are disabled by default right now.

**Comment by karlmdavis** [comment]
As mentioned in the (now updated) description: merging before this is tested, so that others can proceed with integration work against it.

**Pull request merged**